File: project/c_prog/libzahl/src/znot.c
/* See LICENSE file for copyright and license details. */
#include "internals.h"


void
znot(z_t a, z_t b)
{
	size_t bits;

	if (unlikely(zzero(b))) {
		SET_SIGNUM(a, 0);
		return;
	}

	bits = zbits(b);
	a->used = b->used;
	SET_SIGNUM(a, -zsignum(b));

	ZMEM_1OP(a->chars, b->chars, a->used, ~);
	bits = BITS_IN_LAST_CHAR(bits);
	if (bits)
		a->chars[a->used - 1] &= ((zahl_char_t)1 << bits) - 1;

	TRIM_AND_ZERO(a);
}

File: project/llm_from_c_output/libzahl/znot.rs
/* See LICENSE file for copyright and license details. */
use crate::internals::*;
use crate::zahl::*;

pub fn znot(a: &mut Z, b: &Z) {
    let mut bits;

    if unlikely(b.is_zero()) {
        a.set_signum(0);
        return;
    }

    bits = zbits(b);
    a.used = b.used;
    a.set_signum(-b.signum());

    for i in 0..a.used {
        a.chars[i] = !b.chars[i];
    }
    
    bits = bits_in_last_char(bits);
    if bits != 0 {
        a.chars[a.used - 1] &= ((1 as zahl_char_t) << bits) - 1;
    }

    trim_and_zero(a);
}